system/Model.php

```diff
@@ -689,141 +694,139 @@ protected function shouldUpdate($data): bool
+    public function update($id = null, $row = null): bool
     {
         if (! empty($this->tempData['data'])) {
-            if (empty($data)) {
-                $data = $this->tempData['data'];
+            if (empty($row)) {
+                $row = $this->tempData['data'];
             } else {
-                $data = $this->transformDataToArray($data, 'update');
-                $data = array_merge($this->tempData['data'], $data);
+                $row = $this->transformDataToArray($row, 'update');
+                $row = array_merge($this->tempData['data'], $row);
             }
         }
 
         $this->escape   = $this->tempData['escape'] ?? [];
         $this->tempData = [];
 
-        return parent::update($id, $data);
+        return parent::update($id, $row);
     }
 
     /**
      * Takes a class and returns an array of its public and protected
      * properties as an array with raw values.
      *
-     * @param object|string $data
-     * @param bool          $recursive If true, inner entities will be cast as array as well
+     * @param object $object    Object
+     * @param bool   $recursive If true, inner entities will be cast as array as well
      *
-     * @return array|null Array
+     * @return array<string, mixed>
      *
      * @throws ReflectionException
      */
-    protected function objectToRawArray($data, bool $onlyChanged = true, bool $recursive = false): ?array
+    protected function objectToRawArray($object, bool $onlyChanged = true, bool $recursive = false): array
     {
-        $properties = parent::objectToRawArray($data, $onlyChanged);
+        $properties = parent::objectToRawArray($object, $onlyChanged);
 
         $primaryKey = null;
 
-        if ($data instanceof Entity) {
-            $cast = $data->cast();
+        if ($object instanceof Entity) {
+            $cast = $object->cast();
 
             // Disable Entity casting, because raw primary key data is needed for database.
-            $data->cast(false);
+            $object->cast(false);
 
-            $primaryKey = $data->{$this->primaryKey};
+            $primaryKey = $object->{$this->primaryKey};
 
             // Restore Entity casting setting.
-            $data->cast($cast);
-        }
+            $object->cast($cast);
 
-        // Always grab the primary key otherwise updates will fail.
-        if (
-            // @TODO Should use `$data instanceof Entity`.
-            method_exists($data, 'toRawArray')
-            && (
+            // Always grab the primary key otherwise updates will fail.
+            if (
                 ! empty($properties)
                 && ! empty($this->primaryKey)
                 && ! in_array($this->primaryKey, $properties, true)
                 && ! empty($primaryKey)
-            )
-        ) {
-            $properties[$this->primaryKey] = $primaryKey;
+            ) {
+                $properties[$this->primaryKey] = $primaryKey;
+            }
```

Comment:
The primary key may be not filled, which probably will break things for someone. Even the comment before states:
> // Always grab the primary key otherwise updates will fail.

Reply:
Ahh... now I see what you mean. Well, that's sucks... the breaking change actually was made before this. The code used to be:
```php
if (method_exists($data, 'cast')) {
```
instead of the check for the Entity instance.

Reply:
Actually, I'm not sure why we decided to tightly couple everything to the Entity class. We used to give much more freedom in this matter.

Reply:
I meant this is already broken now. So this refactoring will not break it.
Do you think we should fix this (line 803) as a bug?

Reply:
I think we should not be tightly coupled with the Entity class. The path of checking for certain method availability was much more flexible and allowed users to implement their own entity. IMO we shouldn't take this feature away.

Reply:
Okay, then I will fix this.

Reply:
See #8271
To be honest, I'm not sure how to fix it.

- `toRawArray()` should return array that can be inserted into database, but the current CI4 Entity does not.

Reply:
I commented on the linked PR.